protocols/mdns/src/behaviour.rs

```diff
@@ -18,79 +18,34 @@
-use crate::query::MdnsPacket;
-use crate::IPV4_MDNS_MULTICAST_ADDRESS;
-use async_io::{Async, Timer};
+use crate::iface::InterfaceState;
+use crate::MdnsConfig;
+use async_io::Timer;
+use fnv::FnvHashMap;
 use futures::prelude::*;
 use if_watch::{IfEvent, IfWatcher};
 use libp2p_core::connection::ListenerId;
-use libp2p_core::{address_translation, multiaddr::Protocol, Multiaddr, PeerId};
+use libp2p_core::{Multiaddr, PeerId};
 use libp2p_swarm::{
     protocols_handler::DummyProtocolsHandler, NetworkBehaviour, NetworkBehaviourAction,
     PollParameters, ProtocolsHandler,
 };
 use smallvec::SmallVec;
-use socket2::{Domain, Socket, Type};
-use std::{
-    cmp,
-    collections::VecDeque,
-    fmt, io, iter,
-    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr, UdpSocket},
-    pin::Pin,
-    task::Context,
-    task::Poll,
-    time::{Duration, Instant},
-};
-
-/// Configuration for mDNS.
-#[derive(Clone, Debug)]
-pub struct MdnsConfig {
-    /// TTL to use for mdns records.
-    pub ttl: Duration,
-    /// Interval at which to poll the network for new peers. This isn't
-    /// necessary during normal operation but avoids the case that an
-    /// initial packet was lost and not discovering any peers until a new
-    /// peer joins the network. Receiving an mdns packet resets the timer
-    /// preventing unnecessary traffic.
-    pub query_interval: Duration,
-    /// IP address for multicast.
-    pub multicast_addr: IpAddr,
-}
-
-impl Default for MdnsConfig {
-    fn default() -> Self {
-        Self {
-            ttl: Duration::from_secs(6 * 60),
-            query_interval: Duration::from_secs(5 * 60),
-            multicast_addr: *IPV4_MDNS_MULTICAST_ADDRESS,
-        }
-    }
-}
+use std::collections::hash_map::Entry;
+use std::{cmp, fmt, io, net::IpAddr, pin::Pin, task::Context, task::Poll, time::Instant};
 
 /// A `NetworkBehaviour` for mDNS. Automatically discovers peers on the local network and adds
 /// them to the topology.
 #[derive(Debug)]
 pub struct Mdns {
-    /// Main socket for listening.
-    recv_socket: Async<UdpSocket>,
-
-    /// Query socket for making queries.
-    send_socket: Async<UdpSocket>,
+    /// InterfaceState config.
+    config: MdnsConfig,
 
     /// Iface watcher.
     if_watch: IfWatcher,
 
-    /// Buffer used for receiving data from the main socket.
-    /// RFC6762 discourages packets larger than the interface MTU, but allows sizes of up to 9000
-    /// bytes, if it can be ensured that all participating devices can handle such large packets.
-    /// For computers with several interfaces and IP addresses responses can easily reach sizes in
-    /// the range of 3000 bytes, so 4096 seems sensible for now. For more information see
-    /// [rfc6762](https://tools.ietf.org/html/rfc6762#page-46).
-    recv_buffer: [u8; 4096],
-
-    /// Buffers pending to send on the main socket.
-    send_buffer: VecDeque<Vec<u8>>,
+    /// Mdns interface states.
+    iface_states: FnvHashMap<IpAddr, InterfaceState>,
```

Comment:
I understand that using a `FnvHashMap` increases performance here. Though is it worth the additional dependency? How is this a bottleneck in the first place? A lookup should be on the order of 100 ns either way, no?

Reply:
removed it, but it's not really an additional dependency as libp2p-core uses it

Reply:
Thanks. I know. I think it is still worth removing.